File: y15d03/main.go
package main

import (
	"flag"
	"fmt"
	"strings"
)

func main() {
	f := flag.String("f", "", "Input File Name")
	flag.Parse()
	lines := readLines(*f)
	solve(lines)
}

var moveDir = map[string]Vector{"^": {-1, 0}, "v": {1, 0}, "<": {0, -1}, ">": {0, 1}}

func (v Vector) move(m string) Vector {
	o := moveDir[m]
	return v.Add(o)
}

func solve(lines []string) {
	moves := strings.Split(lines[0], "")
	santaPos, bothPos := Vector{0, 0}, []Vector{{0, 0}, {0, 0}}
	visited1, visited2 := make(map[Vector]int), make(map[Vector]int)
	for i, move := range moves {
		santaPos = santaPos.move(move)
		visited1[santaPos] = 1 // for part 1

		bothPos[i%2] = bothPos[i%2].move(move)
		visited2[bothPos[i%2]] = 1 // for part 2
	}
	fmt.Println("Part 1", len(visited1))
	fmt.Println("Part 2", len(visited2))
}

File: y15d03/main.cpp
#include "helper.h"
#include <iostream>
#include <map>
#include <ostream>
#include <set>
#include <vector>
using namespace std;

int main(int argc, char *argv[]) {
  if (argc < 2) {
    cout << "Provide the input file name" << endl;
    return 1;
  }

  const string moves = read_lines(argv[1])[0];
  Position santa_pos;
  const map<char, Position> move_dir = {{'^', Position(-1, 0)},
                                        {'v', Position(1, 0)},
                                        {'<', Position(0, -1)},
                                        {'>', Position(0, 1)}};

  set<Position> houses1 = {santa_pos};

  Position both_pos[] = {Position(), Position()};
  set<Position> houses2 = {both_pos[0]};

  for (size_t i = 0; i < moves.size(); i++) {
    // find direction
    const auto &dir = moves[i];
    auto it = move_dir.find(dir);
    Position other = it->second;

    // Part 1
    santa_pos = santa_pos + other;
    houses1.insert(santa_pos);

    // Part 2
    both_pos[i % 2] = both_pos[i % 2] + other;
    houses2.insert(both_pos[i % 2]);
  }

  cout << "Part 1: " << houses1.size() << endl;
  cout << "Part 2: " << houses2.size() << endl;

  return 0;
}
